doc/cprover-manual/jbmc-user-manual.md

```diff
@@ -0,0 +1,229 @@
+possible inputs (argument values) for which they can be violated:
+```
+core/Example.java function java::core.Example.query:([II)I
+[java::core.Example.query:([II)I.1] line 6 no uncaught exception: SUCCESS
+[java::core.Example.query:([II)I.null-pointer-exception.1] line 9 Null pointer check: SUCCESS
+[java::core.Example.query:([II)I.null-pointer-exception.2] line 12 Null pointer check: SUCCESS
+[java::core.Example.query:([II)I.array-index-out-of-bounds-low.1] line 12 Array index should be >= 0: SUCCESS
+[java::core.Example.query:([II)I.array-index-out-of-bounds-high.1] line 12 Array index should be < length: SUCCESS
+```
+
+\subsection jbmc-manual-unwind Loop unwinding
+
+As in CBMC, JBMC transforms the input program into a state representation,
+and thus needs to unwind loops up to a given value. Without specifying a
+unwind limit, JBMC will not terminate when analysing the following function
+whose state representation is theoretically unbounded (because of the for-loop
+whose number of iterations depends on an input):
+```java
+    public static boolean isPrime(int candidate) {
+        for (int div = 2; div <= candidate / 2; div++) {
+            if (candidate % div == 0) {
+                return false;
+            }
+        }
+        return candidate > 1;
+    }
+```
+To limit the number of times the for-loop is unwound, we use the `--unwind N`
+ options, in which case the following call to JBMC:
+```
+jbmc Example.class --function core.Example.isPrime --unwind 10
+```
+will terminate (with `VERIFICATION SUCCESSFUL`).
+
+
+\subsection jbmc-manual-user-assert User assertions
+
+Elaborating on the example above, users may provide their own assertions, that
+ JBMC will try do refute. Consider the following class:
+ ```java
+1   package core;
+2
+3   public class Example {
+4
+5       public static void doSomething(int input) {
+6           if (input > 1 && input % 2 == 1) {
+7               assert isPrime(input);
+8               // do something
+9           }
+10      }
+11
+12      public static boolean isPrime(int candidate) {
+13          for (int div = 2; div * div <= candidate; div++) {
+14              if (candidate % div == 0) {
+15                  return false;
+16              }
+17          }
+18          return candidate > 1;
+19      }
+20
+21  }
+ ```
+
+On line 7, we check the assertion that all odd numbers greater than 1 are prime.
+To be sure that this always holds, we run JBMC on the example, with a
+reasonable `unwind` value:
+```
+jbmc Example.class --function core.Example.isPrime --unwind 10
+```
+and unsurprisingly JBMC doesn't agree, and prints an assertion failure
+(truncated here for readability):
+```
+ [...doSomething:(I)V.assertion.1] line 7 assertion at file core/Example.java: FAILURE
+```
+
+Rerunning the analysis with the `--trace` option, we can see find out which
+input value JBMC found to trigger the violation:
+```
+INPUT arg0i: 15
```

Comment:
To getter better variable names, compile with `-g` and explain this.

Reply:
Done